frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -631,16 +642,24 @@ pub mod pallet {
 			}
 		}
 
-		fn offchain_worker(n: T::BlockNumber) {
-			// We only run the OCW in the first block of the unsigned phase.
-			if Self::current_phase().is_unsigned_open_at(n) {
-				match Self::try_acquire_offchain_lock(n) {
-					Ok(_) => {
-						let outcome = Self::mine_check_and_submit().map_err(ElectionError::from);
-						log!(info, "mine_check_and_submit execution done: {:?}", outcome);
+		fn offchain_worker(now: T::BlockNumber) {
+			match Self::current_phase() {
+				Phase::Unsigned((true, opened)) if opened == now => {
+					// mine a new solution, cache it, and attempt to submit it
+					let initial_output = Self::try_acquire_offchain_lock(now)
+						.and_then(|_| Self::mine_check_save_submit());
+					log!(info, "initial OCW output at {:?}: {:?}", now, initial_output);
+				}
+				Phase::Unsigned((true, opened)) if opened < now => {
+					if !<QueuedSolution<T>>::exists() {
```

Comment:
https://github.com/paritytech/substrate/pull/8253 (currently draft, not yet ready for review) is intended to address exactly this issue. It lets people challenge a bad solution.

Reply:
what I think is that if the validator has generated one solution offchain he can quickly check that his solution is better or worse that the queued one.
Considering that the validator should have generated a solution when `opened == now`, I think we should be able to quickly check from this generated solution.
Or if he didn't because he generated for a fork. Then we can force validator to generate at one solution per finalized chain maybe.

Currently we do:
* generate one solution (but it might be for a fork) and try submit it.
* if there is no solution on chain after N blocks: retry to generate a solution or take the previously generated one and try submit it.

I wonder why we don't do:
* generate one solution (but it might be for a fork) and try submit it.
* after N blocks: retry to generate a solution or take the previously generated one and try submit it if it has a better score than the on-chain one.

Reply:
My reasoning there is that it's wasteful. We always submit at the first opportunity. That means that the only time a continuous resubmit like that would be valuable is when all of 

- our initial solution wasn't accepted, probably due to network issues / cosmic rays
- someone else submitted a solution which is provably fair (passes PJR check)
- but ours is better (higher score)

This feels to me like it fits at the intersection of low-probability-of-occurring and low-benefit-if-implemented.

Reply:
Hmm actually my thought were confused. We save the solution in a [persistent](https://substrate.dev/rustdocs/v3.0.0/sp_runtime/offchain/storage/struct.StorageValueRef.html#method.persistent) offchain storage. Thus if we generate a solution for the wrong election data snapshot. Our solution is still bad.
The election data snapshot can be different between forks if the fork happened before the last changed vote.

If it happens that we generated the solution for a different snapshot I suggest that the node try to run the election again for the new snapshot and compare the solution score with the on-chain one.

I think nodes should generate a solution for all the election data snapshots.

Reply:
I don't think I properly understand your proposed mitigation. Are you saying that we should be using [`local`](https://substrate.dev/rustdocs/v3.0.0/sp_runtime/offchain/storage/struct.StorageValueRef.html#method.local) instead of `persistent`, because that's documented as fork-aware?

Reply:
> #8253 (currently draft, not yet ready for review) is intended to address exactly this issue. It lets people challenge a bad solution.

This challenge phase is only intended for a signed solution. 

> This feels to me like it fits at the intersection of low-probability-of-occurring and low-benefit-if-implemented.

Yes, it is unlikely for a validator to submit a solution that is relatively a lot worse than ours, so you are right about _low-probability-of-occurring_. What you are (very) wrong about is _low-benefit-if-implemented_. We're talking about multibillion dollar systems and given such an amount at stake, a validator tweaking their offchain worker to produce a shitty solution (due to any malicious reason) is not unreasonable. 

So the initial point is super valid and needs to be addressed. I see two options:

Reply:
1. resubmit if a solution is not queued, or if a solution is queued, and we have once cached, and we can beat the current solution. I am not sure how, but I am worried that this can still be tricked. For example, if somehow all validators start building on a fork in their initial OCW, and don't store anything in the OCW cache, but once malicious validator puts a shitty solution, then we never challenge it. 
2. We fundamentally change the system: We dictate each OCW to at least run `N` (e.g. 5) times, ensuring that the chance of censorship is very little. So each validator will run, with intervals of 5, 5 OCW calls (at blocks [0, 5, 10, 15, 20, 25] after unsigned phase started). In this case we won't need the OCW storage at all, and there's a really good chance that if there's a fork, we recover from it and someone will eventually submit something.

Reply:
Option 2 is a bit of revamp, but I like it a lot. Note that the current proposed solution has never been tested (because to test it we need a fork on the session boundary). 

I think the current option also has a flaw: If the solution that is stored in the OCW cache is _invalid_ or generally unacceptable, we never tweak it and blindly try and re-submit.

Reply:
> This challenge phase is only intended for a signed solution.

Yes, true. I'd missed that somehow.

---

ea3fc2908 implements option 1. I see that you like option 2, but I was already working on this when that message came in. If you like, I'll revert it and implement option 2 instead.

--- 

> If the solution that is stored in the OCW cache is invalid

Right, I was thinking about that also. I think cc70d3767 at least partially solves it, but it may be incomplete; I'm not sure there's a point in having a cache at all, in that case.

Reply:
> Right, I was thinking about that also. I think cc70d37 at least partially solves it, but it may be incomplete; I'm not sure there's a point in having a cache at all, in that case.

Yeah this is the right direction, but we have to double check when we get `PoolSubmissionFailed` as result. If it does call `validate_unsigned` and fail immediately if it fails, then this is very good, else, not much helpful.

Reply:
> I don't think I properly understand your proposed mitigation. Are you saying that we should be using [`local`](https://substrate.dev/rustdocs/v3.0.0/sp_runtime/offchain/storage/struct.StorageValueRef.html#method.local) instead of `persistent`, because that's documented as fork-aware?

Yes and no in case we do the option1 from Kian, we need either to make use of fork-aware storage, or store the validator/nominators sets on which we computed the solution and compare with the current one.
Or test that our solution is valid, if it is not valid then it is outdated and we need to compute one again.

If we can easily afford computing 5 solutions then option2 from Kian seems ok.

For me in option1 we should be able to know if our solution is outdated because of fork. So it is enough to mitigate this situations.
for option2 that also allow us to recompute solutions in case our score is bad for some reason. Seems easier but not strictly needed.

Reply:
> ea3fc29 implements option 1. I see that you like option 2, but I was already working on this when that message came in. If you like, I'll revert it and implement option 2 instead.

Hmm what I meant was more like this: 

```
if !<QueuedSolution<T>>::exists || restore_solution::<T>().map_or(true, |call| call.score.is_better(QueuedSolution<T>::get().score)) {
// syntax is a bit simplified, will need a rather ugly match on call. 
}
```

Reply:
I've never used `local`, but based on the documentation it is exactly what we need. If a solution is computed in a fork, then it should not be visible once we re-org out of that fork. 

What I would be curious in this case is what happens to that data? if it is not wiped, we should make sure it is cleaned at some point. 

In either case, it might be a good idea to clear all OCW storage once `ElectionFinalized`.

Reply:
> Hmm what I meant was more like this

I think the only difference between what you wrote there and what ea3fc29 implements is when we perform the queued solution check. Logically, it still only submits if the queued solution does not exist or if the local solution has a better score than what has been queued on-chain.

Reply:
> Logically, it still only submits if the queued solution does not exist or if the local solution has a better score than what has been queued on-chain.

Yeah my bad, didn't check the rest of the code. Also good call to add `maybe` to the function name.